### ISSUE TITLE
Possible bug with unused variables

### ISSUE THREAD

**Issue by thr0wn** [Bitcoin-com/cashscript#83]
How to reproduce:
https://codesandbox.io/s/blazing-bash-qiwdo?file=/src/App.js

Basically, SymbolTableTraversal does not recognize expressions like `foo * bar` as references to both `foo` and `bar` variables.

Maybe something related to [this line](https://github.com/Bitcoin-com/cashscript/blob/b08e55a78b2e48a6e4dd6100c29758e845eb7fdb/packages/cashc/src/ast/SymbolTable.ts#L75)?

**Comment by p0o** [comment]
Confirmed. A simple (dumb) workaround would be to put `bar = bar;` before require

**Comment by rkalis** [comment]
So the real issue here is that you're trying to use multiplication, which is unsupported by the BCH Script engine. Somehow the compiler still tries to do some error recovery and continues compilation (ignoring the `* bar` part). Since `* bar` gets ignored by the compiler, it then complains that `bar` is not used.

I'm going to look if I can just make the compilar fail early (instead of trying to do the error recovery).

**Comment by p0o** [comment]
Ahh true I missed the fact that it's multiplying. But what's the idea with this error recovery? I don't think it's a good thing for compiler to ignore a part of the business logic and compile the contract. Someone using that might think the contract contains some logic that is actually ignored.

**Comment by rkalis** [comment]
Yeah I agree. It's the default behaviour of ANTLR (the compiler tool that CashScript uses), but I'll update it so that it doesn't do that any more 😅

**Comment by rkalis** [comment]
Just published `v0.5.6` which now fails early on lexing/parsing errors (rather than doing error recovery).

The error messages are a bit generic since they're generated by ANTLR, but they're clear enough to be actionable.

```
$ cashc ./bug.cash
line 7:21 token recognition error at: '*'
```

Thanks for bringing this to my attention :muscle:

**Comment by rkalis** [comment]
Actually just noticed that it gives an empty error if you forget a semi-colon 😓

**Comment by rkalis** [comment]
Alright, just published a new release `v0.5.7` that fixes the missing error messages on certain error (e.g. the semi-colon). Thanks again for the report!

**Comment by thr0wn** [comment]
Thanks, @rkalis!
